tests/wasmfs/wasmfs_stat.c

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2021 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+// FIXME: Individual test to verify fstat in isolation. May get merged with
+// others later.
+
+int main() {
+  // Attempt to call fstat on an invalid fd.
+  errno = 0;
+  struct stat invalid;
+  int result = fstat(-1, &invalid);
+  assert(result == -1);
+  assert(errno == EBADF);
+
+  // Test opening a file and calling fstat.
+  struct stat file;
+  int fd = open("/dev/stdout/", O_RDONLY);
+  fstat(fd, &file);
+  assert(fd >= 0);
+
+  off_t fileInode = file.st_ino;
+
+  assert(file.st_size == 0);
+#ifdef WASMFS
+  assert((file.st_mode & S_IFMT) == S_IFREG);
+#else
+  assert((file.st_mode & S_IFMT) == S_IFCHR);
+#endif
+  assert(file.st_dev);
+  assert(file.st_nlink);
+  assert(file.st_uid == 0);
+  assert(file.st_gid == 0);
+  assert(file.st_rdev);
+  assert(file.st_blocks == 0);
+  assert(file.st_blksize == 4096);
+#ifdef WASMFS
+  assert(file.st_atim.tv_sec == 0);
+  assert(file.st_mtim.tv_sec == 0);
+  assert(file.st_ctim.tv_sec == 0);
+#endif
+
+  close(fd);
+
+  // Check to see if the previous inode number matches.
+  int newfd = open("/dev/stdout/", O_RDONLY);
+  struct stat newFile;
+  fstat(newfd, &newFile);
+
+  assert(fileInode == newFile.st_ino);
+  close(newfd);
+
+  // Test opening a directory and calling fstat.
+  struct stat directory;
+  int fd2 = open("/dev", O_RDONLY | O_DIRECTORY);
+  fstat(fd2, &directory);
+  assert(fd2 >= 0);
+
+  off_t dirInode = directory.st_ino;
+
+  assert(directory.st_size == 4096);
+  assert((directory.st_mode & S_IFMT) == S_IFDIR);
+  assert(directory.st_dev);
+  assert(directory.st_nlink);
+  assert(directory.st_uid == 0);
+  assert(directory.st_gid == 0);
+#ifdef WASMFS
+  assert(directory.st_rdev);
+  // blkcnt_t  st_blocks;      /* Number of 512B blocks allocated */
+  assert(directory.st_blocks == 8);
```

Comment:
What does the JS system do for `st_blocks`?

Reply:
https://github.com/emscripten-core/emscripten/blob/7712529643c74eb416d515ba94b832f71ec11e4e/src/library_memfs.js#L166-L169